File: service/src/evidencegraph_service/b64url.py
"""Shared unpadded-Base64URL (RFC 4648 SS5) helpers.

The one text encoding this codebase uses anywhere secret-adjacent bytes
cross a text-based channel: the session-token format itself (`config.py`,
S1-03), the private-channel startup secret (`protocol.py`, S1-04), and the
D-025 challenge nonce/response (`challenge.py`, S1-04) all share this exact
alphabet and this exact decoder -- one implementation, so "what counts as
valid Base64URL here" can never quietly drift between them.
"""

from __future__ import annotations

import base64
import re

#: `A-Z a-z 0-9 - _`, no `=` padding -- see `config.py`'s
#: `is_valid_session_token_format` docstring for the full rationale.
TOKEN_PATTERN = re.compile(r"^[A-Za-z0-9_-]+$")

#: Generous upper bound for any single Base64URL-encoded field this
#: codebase decodes (secrets/nonces/responses are ~32-64 raw bytes, i.e.
#: ~43-86 encoded chars) -- exists only to bound parsing work on a value
#: that hasn't been validated yet, per the private-channel/challenge
#: "messages are bounded in size and strictly parsed" requirement.
MAX_FIELD_LENGTH = 256


class DecodeError(Exception):
    """Malformed or oversized Base64URL text. The message is always one of
    a small set of fixed, bounded, non-secret category strings (the field
    name plus a reason) -- never the offending value itself."""


def encode(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")


def decode(value: object, *, field: str) -> bytes:
    """Strictly parses `value` as unpadded Base64URL. Rejects anything
    that isn't a non-empty, bounded-length string matching `TOKEN_PATTERN`
    *before* ever calling into the decoder -- never a best-effort decode
    of untrusted input."""
    if not isinstance(value, str) or not value or len(value) > MAX_FIELD_LENGTH:
        raise DecodeError(f"{field}_invalid_length")
    if not TOKEN_PATTERN.fullmatch(value):
        raise DecodeError(f"{field}_invalid_encoding")
    padded = value + "=" * (-len(value) % 4)
    try:
        return base64.urlsafe_b64decode(padded)
    except ValueError as exc:  # binascii.Error subclasses ValueError
        raise DecodeError(f"{field}_invalid_encoding") from exc

File: app/src-tauri/src/supervisor/b64url.rs
//! Unpadded Base64URL (RFC 4648 §5) -- mirrors
//! `service/src/evidencegraph_service/b64url.py` exactly. The one text
//! encoding the private-channel/challenge wire protocol uses for
//! secret-bearing byte fields (startup secret, nonce, challenge response,
//! session token).

use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use base64::Engine as _;

pub fn encode(raw: &[u8]) -> String {
    URL_SAFE_NO_PAD.encode(raw)
}

pub fn decode(value: &str) -> Result<Vec<u8>, base64::DecodeError> {
    URL_SAFE_NO_PAD.decode(value)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trips() {
        let raw: Vec<u8> = (0u8..=255).collect();
        assert_eq!(decode(&encode(&raw)).unwrap(), raw);
    }

    #[test]
    fn never_contains_padding() {
        for len in 1..40 {
            let raw = vec![0xABu8; len];
            assert!(!encode(&raw).contains('='));
        }
    }

    #[test]
    fn matches_a_known_python_encoded_value() {
        // `python -c "import base64; print(base64.urlsafe_b64encode(bytes(range(4))).decode())"`
        // -> "AAECAw==" -> stripped of padding: "AAECAw"
        assert_eq!(encode(&[0, 1, 2, 3]), "AAECAw");
    }
}
